### PR TITLE
Choleksy Factorization miniapp: migration to random values input

### LINKED ISSUE #116: Miniapp for Cholesky Factorization
Implement a miniapp useful for benchmarking of the Cholesky Factorization (using the one from the prototype as reference).

### PR DESCRIPTION
**Pull request by albestro**
Close #116, #117 

Migration from analytical to random values input for Cholesky Factorization Miniapp.

This opens to the benchmarks since it does not suffer of the denormalized values performance issue. Indeed, a revision of this work in progress has been used to produce benchmark results reported in #164.

### CODE REVIEW

**Comment by albestro** [comment]
It still has some `TODO` inside and work in progress, but moreover it has the limitation that it works just with matrices which size is multiple of tile size.

File: miniapp/miniapp_cholesky.cpp [hunk 1]
```diff
@@ -180,14 +194,270 @@ int main(int argc, char** argv) {
...
+      matrix(tile_wrt_global).then(hpx::util::unwrapping([tl_index](auto&& tile) {
+        for (int j = 0; j < tile.size().cols(); ++j)
+          for (int i = 0; i < tile.size().rows(); ++i) {
+            GlobalElementIndex element_wrt_global{tl_index.row() + i, tl_index.col() + j};
+            if (element_wrt_global.row() < element_wrt_global.col())
+              tile(TileElementIndex{i, j}) = 0;
+          }
+      }));
+    }
+  }
```
Comment (rasolca) [inline miniapp/miniapp_cholesky.cpp:287]: I would simplify this in the following way:
```
loop over i_local
  convert i_local to i_global
  j_global = i_global
  if j_global in this rank:
    convert j_global to j_local
    schedule a task that performs lapack laset with m=row_tile_size-1 n=col_tile_size-1 and ptr=tile.ptr(0,1)
```

File: miniapp/miniapp_cholesky.cpp [hunk 2]
```diff
@@ -180,14 +194,270 @@ int main(int argc, char** argv) {
...
+///
+/// 1. Compute the max norm of the original matrix
+/// 2. Compute the difference between the original and the computed matrix using the factorization
+/// 3. Compute the max norm of the difference
+/// 4. Evaluate the correctness of the result using the ratio between the two matrix max norms
+///
+/// Prints a message with the ratio and a note about the error:
+/// "":        check ok
+/// "ERROR":   error is high, there is an error in the factorization
+/// "WARNING": error is slightly high, there should be an error in the factorization
```
Comment (rasolca) [inline miniapp/miniapp_cholesky.cpp:420]:
```suggestion
/// "WARNING": error is slightly high, there can be an error in the factorization
```

**Comment by albestro** [comment]
Close #160 
This does not depend anymore on code from `DLAF_test`

**Comment by albestro** [comment]
Last refinements undergoing I will complete soon.

Apart from them, and before opening this PR for a complete review, there are few aspects which I'd like to highlight (they may be addressed separately in the future, in case of needs).

1. About matrix initialization in the miniapp: do we prefer to initialize it with random values and then copy it when needed, or do we prefer to initialize it randomly always with the same seed?
2. There may be room for lowering the memory usage of `cholesky_diff` on `partial_result`, but it may complicate a bit the logic for reducing them over different ranks.
3. @teonnik pointed out a potential problem with wrapping `std::vector` (see issue #221)

I'm willing to get a feedback from you (@rasolca @teonnik @Sely85)

**Comment by albestro** [comment]
bors try

**Comment by albestro** [comment]
I merged the new master in which it is available `norm(max)` for the specific variant I need in this PR.

> [...] there are few aspects which I'd like to highlight (they may be addressed separately in the future, in case of needs).
> 
>     1. About matrix initialization in the miniapp: do we prefer to initialize it with random values and then copy it when needed, or do we prefer to initialize it randomly always with the same seed?
> 
>     2. There may be room for lowering the memory usage of `cholesky_diff` on `partial_result`, but it may complicate a bit the logic for reducing them over different ranks.
> 
>     3. @teonnik pointed out a potential problem with wrapping `std::vector` (see issue #221)
> 
> 

I'm willing to get a feedback from you @rasolca @teonnik @Sely85

**Comment by teonnik** [comment]
> 1. About matrix initialization in the miniapp: do we prefer to initialize it with random values and then copy it when needed, or do we prefer to initialize it randomly always with the same seed?

Assuming the values in the matrix don't influence the performance of Cholesky much (which I'm not 100% sure about), I don't mind leaving it as it is for now... However setting random values is generally somewhat slow, it seems preferable to set them once at the start.  

> 2. There may be room for lowering the memory usage of `cholesky_diff` on `partial_result`, but it may complicate a bit the logic for reducing them over different ranks.

I think it's OK to leave that for a separate PR. I presume this is a part of the code will not be included in benchmarks.

> 3. @teonnik pointed out a potential problem with wrapping `std::vector` (see issue #221)

This is not an urgent issue, I'm OK with leaving things as they are for now in order to unblock other PRs.

**Comment by albestro** [comment]
> > 1. About matrix initialization in the miniapp: do we prefer to initialize it with random values and then copy it when needed, or do we prefer to initialize it randomly always with the same seed?
> 
> Assuming the values in the matrix don't influence the performance of Cholesky much (which I'm not 100% sure about), I don't mind leaving it as it is for now... However setting random values is generally somewhat slow, it seems preferable to set them once at the start.

I've just changed this: now the miniapp initializes the matrix reference with random values (just one time) and then at each run it copies the values from it (once for the running the algorithm and an optional one if the results are checked).

> > 1. There may be room for lowering the memory usage of `cholesky_diff` on `partial_result`, but it may complicate a bit the logic for reducing them over different ranks.
> 
> I think it's OK to leave that for a separate PR. I presume this is a part of the code will not be included in benchmarks.

Yes, it does not affect the benchmark results since it is in the part for checking results.

**Comment by albestro** [comment]
bors try

**Comment by teonnik** [comment]
Note that this PR also fixes build issues when DLAF_WITH_TEST=OFF as the DLAF_test target used in the miniapp is no longer needed.

**Comment by albestro** [comment]
> Note that this PR also fixes build issues when DLAF_WITH_TEST=OFF as the DLAF_test target used in the miniapp is no longer needed.

Yes! The issue had been already linked for being closed 😉

**Comment by Sely85** [comment]
>> 1. About matrix initialization in the miniapp: do we prefer to initialize it with random values and then copy it when needed, or do we prefer to initialize it randomly always with the same seed?
>>
> I've just changed this: now the miniapp initializes the matrix reference with random values (just one time) and then at each run it copies the values from it (once for the running the algorithm and an optional one if the results are checked).

This looks great! I do prefer the first option as well.


>> 2. There may be room for lowering the memory usage of cholesky_diff on partial_result, but it may complicate a bit the logic for reducing them over different ranks.
>>
> Yes, it does not affect the benchmark results since it is in the part for checking results.

Perfect: if it does not affect benchmark results, I think it can be left as it is to avoid complicating the logic. 


>> 3. @teonnik pointed out a potential problem with wrapping std::vector (see issue #221)
>>
> This is not an urgent issue, I'm OK with leaving things as they are for now in order to unblock other PRs.

If it's not urgent, I agree not to modify anything for now.

**Pull request merged**